cbmc_viewer/memopt.py

```diff
@@ -0,0 +1,204 @@
+        )
+
+    def dump(self, filename=None, outdir=None):
+        """Write memop metrics to a file or stdout."""
+
+        util.dump(self, filename, outdir)
+
+    def render_report(self, filename=None, outdir=None):
+        """Write the memop summary to a file rendered as html."""
+
+        memop_html = templates.render_memop_summary(self.summary)
+        util.dump(memop_html, filename or "memop.html", outdir)
+
+################################################################
+# Json key tags to access elements in cbmc json output
+
+JSON_FUNCTIONS_KEY = 'functions'
+JSON_BODY_AVAIL_KEY = 'isBodyAvailable'
+JSON_INSTRUCTIONS_KEY = 'instructions'
+JSON_INSTRUCTION_ID_KEY = 'instructionId'
+JSON_OPERANDS_KEY = 'operands'
+JSON_FUNCTION_INSTR_KEY = 'instruction'
+JSON_SOURCE_LOC_KEY = 'sourceLocation'
+
+# Json keys for CBMC operands
+JSON_ID_KEY = 'id'
+JSON_NAMED_SUB_KEY = 'namedSub'
+JSON_IDENTIFIER_KEY = 'identifier'
+
+# List of memory operations
+list_mem_func = ["memcmp","memcpy","memmove","memset","memccpy","memchr"]
+
+# Example cbmc json output
+# [
+#   {
+#     "program": "CBMC 5.13.0 ..."
+#   },
+#   ...
+#   {
+#     "functions": [
+#       {
+#         "instructions": [
+#           {
+#             "guard": {
+#               "id": "nil"
+#             },
+#             "instruction": "... struct.c line 23...\n...memcmp...\n",
+#             "instructionId": "FUNCTION_CALL",
+#             "operands": [
+#               ...
+#               {
+#                 "id": "symbol",
+#                 "namedSub": {
+#                   "identifier": {
+#                     "id": "memcmp"
+#                   },
+#                   ...
+#                 }
+#               },
+#               ...
+#             ],
+#             "sourceLocation": {
+#               "file": "struct.c",
+#               "function": "main",
+#               "line": "23",
+#               "workingDirectory": "..."
+#             }
+#           },
+#           ...
+#         ],
+#         "isBodyAvailable": true,
+#         ...
+#       }
+#     ]
+#   }
+# ]
+
+################################################################
+# Utility functions to generate memory op call summary
+
+def is_memop_call(operands):
+    for item in operands:
+        if item[JSON_ID_KEY] == "symbol":
+            func_name = \
+                item[JSON_NAMED_SUB_KEY][JSON_IDENTIFIER_KEY][JSON_ID_KEY]
+            return any([True for func in list_mem_func if func in func_name])
+
+    return False
+
+def get_func_call(instr):
+    # instr is of the form "... struct.c line 23...\n...memcmp...\n"
+    # str_list[0] is location, str_list[1] is the call
+    str_list = instr.split('\n')
+    func_call = str_list[1].strip()
+    return func_call
+
+def get_memop_metrics(json_file, srcdir):
+    summary = {
+        'list': [],
+        'count': 0
+    }
+
+    json_data = parse.parse_json_file(json_file)
+    for item in json_data:
```

Comment:
Maybe set functions = [] before this loop so functions is guaranteed to be defined for the next loop

Reply:
Fixed.